ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/worldstate/Pruner.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright 2019 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.worldstate;
+
+import tech.pegasys.pantheon.ethereum.chain.BlockAddedEvent;
+import tech.pegasys.pantheon.ethereum.chain.Blockchain;
+import tech.pegasys.pantheon.ethereum.core.BlockHeader;
+import tech.pegasys.pantheon.ethereum.core.Hash;
+
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class Pruner {
+  private static final Logger LOG = LogManager.getLogger();
+
+  private final MarkSweepPruner pruningStrategy;
+  private final Blockchain blockchain;
+  private final ExecutorService executor;
+  private final long retentionPeriodInBlocks;
+  private final AtomicReference<State> state = new AtomicReference<>(State.IDLE);
+  private volatile long markBlockNumber = 0;
+  private volatile BlockHeader markedBlockHeader;
+  private long transientForkThreshold;
+
+  public Pruner(
+      final MarkSweepPruner pruningStrategy,
+      final Blockchain blockchain,
+      final long retentionPeriodInBlocks,
+      final long transientForkThreshold) {
+    this.pruningStrategy = pruningStrategy;
+    this.blockchain = blockchain;
+    this.executor =
+        Executors.newSingleThreadExecutor(
+            new ThreadFactoryBuilder()
+                .setDaemon(true)
+                .setPriority(Thread.MIN_PRIORITY)
+                .setNameFormat("StatePruning-%d")
+                .build());
+    this.retentionPeriodInBlocks = retentionPeriodInBlocks;
+    this.transientForkThreshold = transientForkThreshold;
+  }
+
+  public void start() {
+    blockchain.observeBlockAdded((event, blockchain) -> handleNewBlock(event));
+  }
+
+  public void stop() {
+    executor.shutdownNow();
+  }
+
+  private void handleNewBlock(final BlockAddedEvent event) {
+    if (!event.isNewCanonicalHead()) {
+      return;
+    }
+    final BlockHeader header = event.getBlock().getHeader();
+    if (state.compareAndSet(State.IDLE, State.TRANSIENT_FORK_OUTLIVING)) {
+      pruningStrategy
+          .prepare(); // implications for calling prepare here when blocks can come from multiple
+      // threads?
```

Comment:
There can only be one thread at a time importing a block, and new block notifications are sent as part of that process, so only ever one thread at a time in this method but which thread it is may change (so we need to ensure changes are visible cross-thread).  So we'll need to make `nodeAddedListenerId` volatile in `MarkSweepPruner` but otherwise it will be safe.